### PR TITLE
Higher precision for floats in range [0,1)

### PR DESCRIPTION
**Pull request by pitdicker**
First a disclaimer :).
With all the discussions happening around rand, I am not sure where to open a pull request.
Also it is a bit difficult for me to explain what I am doing here...

This commit changes the way floats are created with the distribution `Uniform01`.
This is not a method I have found anywhere else, but loosely inspired by
https://readings.owlfolio.org/2007/generating-pseudorandom-floating-point-values/ .

With this method floats have a much better precision near zero (9 extra bits of precision for f32, and 12 extra bits for f64).

It seems to have almost no influence on the benchmarks:
before:
`test distr_baseline          ... bench:       1,616 ns/iter (+/- 68) = 4950 MB/s`
after:
`test distr_baseline          ... bench:       1,621 ns/iter (+/- 30) = 4935 MB/s`

### CODE REVIEW

**Comment by dhardy** [comment]
Nice. I haven't had the time today but I'll take a look and merge.

I haven't focussed on changes like this since getting the main architecture right seemed like a priority, but this is definitely worth bringing up and mentioning on the main RFC thread.

**Comment by pitdicker** [comment]
Thank you, and there is no hurry for me.

I am now looking what the consequences are for other parts of rand that use this.

`range` for example looses most of the precision again if one of it's numbers is negative. I have fixed this, but not tested en benchmarked yet.
`open01` seems to rely on some precision also...

Tomorrow I will try to write something on the RFC tread.

**Comment by pitdicker** [comment]
A quick update:
`open01`, `closed01` and `range` now preserve the extra bits of precision near 0.0.

The benchmark `distr_range2_float` has regressed 40%. Moving the code conversion from uint to float out from `closed01`, `uniform01` and `uniform11` fixes this. Probably some problem with inlining?

This is a good chance to simplify the code to use constants, and wrap it in a macro so there is no duplication between f32 and f64.

Which is a long way of saying, please wait a few more days before reviewing :)

**Comment by pitdicker** [comment]
It should be ok now.

**Comment by pitdicker** [comment]
The code in `ziggurat` in `distributions` uses it's own method of converting a random int to a float.
It only uses as many bits of precision as bit in the floats fraction, and uses the rest to pick a row from the ziggurat. It is perfect to use the previous code from `Uniform01` here. The `distribution` benchmarks improve by 5~7%.

This got me thinking that it may be best to separate out the int to float conversions to an other file. It does not really fit in `uniform.rs`.

I'll work on it some more...

**Comment by dhardy** [comment]
Sure, putting the float conversions in a separate file makes sense (`uniform_float.rs` or `float.rs` or some such).

I'm not so sure about making `uniform01` etc. member functions on `u32` / `u64`. It may be useful to make these public, sure, but in general it seems a weird operation to make a member function. Well, I suppose they are only available when people `use FloatConversions` so maybe the design is fine.

**Comment by pitdicker** [comment]
You are right that making them member functions is not ideal. But I don't know enough rust yet for something better. Do you have a suggestion, or shall I just try a bit more?

**Comment by dhardy** [comment]
Simple functions (generic or with separate names for each type) would work. But I don't know if this would be better. As noted in my edit above, using member functions is probably fine actually.

**Comment by pitdicker** [comment]
Moving the functions to a separate file took care of them being public, so I was lucky :-).

I have tried to clean up the commit history.

Some benchmarks:
Before:
```
test distr_baseline          ... bench:       1,620 ns/iter (+/- 17) = 4938 MB/s
test distr_exp               ... bench:       6,286 ns/iter (+/- 131) = 1272 MB/s
test distr_gamma_large_shape ... bench:      18,190 ns/iter (+/- 317) = 439 MB/s
test distr_gamma_small_shape ... bench:      21,233 ns/iter (+/- 427) = 376 MB/s
test distr_log_normal        ... bench:       6,650 ns/iter (+/- 94) = 1203 MB/s
test distr_normal            ... bench:       6,651 ns/iter (+/- 99) = 1202 MB/s
test distr_range2_float      ... bench:       1,620 ns/iter (+/- 17) = 4938 MB/s
test distr_range2_int        ... bench:       3,102 ns/iter (+/- 16) = 2578 MB/s
test distr_range_float       ... bench:       1,620 ns/iter (+/- 17) = 4938 MB/s
test distr_range_int         ... bench:       3,107 ns/iter (+/- 12) = 2574 MB/s
```

After:
```
test distr_baseline          ... bench:       1,617 ns/iter (+/- 15) = 4947 MB/s
test distr_exp               ... bench:       6,090 ns/iter (+/- 121) = 1313 MB/s
test distr_gamma_large_shape ... bench:      17,419 ns/iter (+/- 215) = 459 MB/s
test distr_gamma_small_shape ... bench:      20,304 ns/iter (+/- 259) = 394 MB/s
test distr_log_normal        ... bench:       5,849 ns/iter (+/- 64) = 1367 MB/s
test distr_normal            ... bench:       5,854 ns/iter (+/- 65) = 1366 MB/s
test distr_range2_float      ... bench:       1,620 ns/iter (+/- 7) = 4938 MB/s
test distr_range2_int        ... bench:       3,098 ns/iter (+/- 12) = 2582 MB/s
test distr_range_float       ... bench:       1,618 ns/iter (+/- 8) = 4944 MB/s
test distr_range_int         ... bench:       3,100 ns/iter (+/- 11) = 2580 MB/s
```

The benchmarks that depend on the ziggurat method have all improved, `normal` and `log_normal` by 12%, `gamma` by 4.5%, and `exp` by 3%. The benchmark `distr_range2_float` does not regress, thanks to splitting the float conversion functions off from those also generating the random integers.

Thanks for the encouragement until now!

**Comment by pitdicker** [comment]
I just learned the `gamma` and derived distributions do not use the ziggurat method. That makes me wonder why they have improved. Maybe the new code for `Open01` is what makes them slightly faster...

**Comment by pitdicker** [comment]
The conversion functions are now moved to a different module. The new names make it clearer what is going on, so that is a plus :-)

File: src/utils.rs [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+//! some range.
+
+use core::mem;
+
+trait FloatFromBits {
+    type F;
+
+    /// Helper function to combine the fraction, exponent and sign into a float.
+    /// `self` includes the bits from the fraction and sign.
+    fn combine_float(self, exponent: i32) -> Self::F;
```
Comment (dhardy) [inline src/utils.rs]: Maybe this should be called something like `binor_exp`? (I would suggest `set_exp` but if the exponent bits aren't currently zero it wouldn't do that.) Also the trait name: maybe `FloatBitOps`?

File: src/distributions/uniform.rs [hunk 2]
```diff
@@ -187,68 +188,51 @@ impl Distribution<bool> for Uniform {
...
+                // in the distribution. That is why, instead of a closed range,
+                // we actually sample from the half-open range (0,1].
+                //
+                // Floating-point numbers have more precision near zero. This
+                // means for a f32 that only 1 in 2^32 samples will give exactly
+                // 0.0. But 1 in 2^23 will give exactly 1.0 due to rounding.
+                // Because the chance to sample 0.0 is so low, this half-open
+                // range is a very good appoximation of a closed range.
+                let x = $next_u(rng);
+                x.open_closed01()
```
Comment (dhardy) [inline src/distributions/uniform.rs:215]: You give a good rationale for implementing `[0, 1]` sampling via `(0, 1]` instead. I wonder what happens with your extra precision sampling under common transformations. E.g. if a user samples `x` via this function, then what's the actual range of `x + 1.0` — probably exactly `[1, 2]` since the loss of precision away from 0 forces rounding? Would `x - 1.0` have range `[-1, 0]`?

File: src/distributions/range2.rs [hunk 3]
```diff
@@ -181,37 +182,116 @@ range_int_impl! { usize, usize }
...
+                    }
+                } else if -low <= high {
+                    RangeFloat::PosAndNeg {
+                        cutoff: low,
+                        scale: (high + low) / -2.0 + low,
+                    }
+                } else {
+                    RangeFloat::NegAndPos {
+                        cutoff: high,
+                        scale: (high + low) / 2.0 - high,
```
Comment (dhardy) [inline src/distributions/range2.rs:249]: It seems the half-open `[low, high)` range works well here, doesn't it? Is a closed version, `[low, high]` feasible? Where `low` is close to zero, the same approximation trick, `(low, high]` may be okay (and equiv for high < 0), but for ranges spanning 0 there's no good solution?

The other question is whether this actually matters much in practice... many applications don't need so much precision in the first place, but maybe some do?

File: src/distributions/mod.rs [hunk 4]
```diff
@@ -174,13 +174,19 @@ fn ziggurat<R: Rng+?Sized, P, Z>(
         // this may be slower than it would be otherwise.)
         // FIXME: investigate/optimise for the above.
```
Comment (dhardy) [inline src/distributions/mod.rs]: I think you can remove this comment; we're already removing the `next_f64` function from `Rng`.

File: src/distributions/mod.rs [hunk 5]
```diff
@@ -174,13 +174,19 @@ fn ziggurat<R: Rng+?Sized, P, Z>(
...
         let bits: u64 = uniform(rng);
-        let i = (bits & 0xff) as usize;
-        let f = (bits >> 11) as f64 / SCALE;
 
         // u is either U(-1, 1) or U(0, 1) depending on if this is a
         // symmetric distribution or not.
-        let u = if symmetric {2.0 * f - 1.0} else {f};
-        let x = u * x_tab[i];
+        // FIXME: the distribution is not open, but half-open.
+        //        Can that cause problems or a bias?
```
Comment (dhardy) [inline src/distributions/mod.rs]: I have no idea, sorry. But would rejection sampling slow it down much anyway, in practice? Maybe it's not significant enough to bother anyway.

File: src/distributions/mod.rs [hunk 6]
```diff
@@ -174,13 +174,19 @@ fn ziggurat<R: Rng+?Sized, P, Z>(
...
+        let f = if symmetric {
+                    bits.closed_open11_fixed()
+                } else {
+                    bits.closed_open01_fixed()
```
Comment (dhardy) [inline src/distributions/mod.rs]: Maybe documentation for these methods should make it very clear which bits are not used?

**Review by dhardy — APPROVED** [review]
With a few minor changes I'd love to merge this, if you're finished tweaking of course (otherwise better to wait; there's no rush for this since the trait stuff is still undecided).

File: src/distributions/mod.rs (hunk 5 above)
Comment (pitdicker) [reply src/distributions/mod.rs]: I have been reading up on the ziggurat method. There even is an improved version from 2015, that should be faster. But I added the note to make sure it didn't get lost

File: src/distributions/range2.rs (hunk 3 above)
Comment (pitdicker) [reply src/distributions/range2.rs:249]: I have spend some hours trying to get the rounding at the edge of the ranges right, but without much success yet... So all the stuff about open/closed is theoretically nice, but not yet the reality :-(. That is one of the reasons I was not sure in the other pr if it made sense to expose a closed range in the treat, because it does not work really out for floats.

But I will give my test for de edge cases a few more tries. You asked above, if I am done tweaking. Maybe not yet completely ;-)

File: src/distributions/uniform.rs (hunk 2 above)
Comment (pitdicker) [reply src/distributions/uniform.rs:215]: Yes, scaling should mostly preserve the precision, but addition forces rounding. I hope the extra precision for `Open01` does not break anyone's code. But the most important reason to use `Open01` is to not get 0.0, for log, division, etc.. I don't imagine much problems in those cases.

`x - 1.0` should have `[-1, 0]` as range. I didn't think about that! But it hurts a bit now that I am focused on precision ;-)

File: src/utils.rs (hunk 1 above)
Comment (pitdicker) [reply src/utils.rs]: It is just a helper function, but you are better at naming! I will change it

File: src/distributions/uniform.rs (hunk 2 above)
Comment (dhardy) [reply src/distributions/uniform.rs:215]: I'm wondering now whether we should permanently document `Closed01` as having range `(0, 1]`, or have another distribution like `OpenClosed01` as well, or instead of `Closed01`. Because not generating 0 may be a useful property.

Of course it may be surprising that `OpenClosed01 + x` will often have range `[x, x+1]` due to rounding, but there's not much we can do about that (not knowing in general how large x is).

File: src/distributions/range2.rs (hunk 3 above)
Comment (dhardy) [reply src/distributions/range2.rs:249]: This is also a good rationale for not worrying too much about which open/closed distributions we have. I checked the "pure math" version of these distributions gives the right open/closed distributions, but what happens in practice due to rounding is another matter. :-/

File: src/distributions/uniform.rs (hunk 2 above)
Comment (pitdicker) [reply src/distributions/uniform.rs:215]: I would not hurry with adding more range distribution with a different closed/open story... We already have maybe to many to choose from. If someone doesn't want to generate 0.0, he can use `Open01`. And otherwise there is now the conversion function `open_closed01`. So I would not do add it right now, but who knows what the future brings?

**Comment by pitdicker** [comment]
Someday I will learn how to rebase properly...

I have made the changes you suggested.
I am done tweaking this code. Maybe in the future I will have another try at the rounding with ranges.

**Comment by dhardy** [comment]
Great, thanks. Something like `git rebase -i HEAD~5` is extremely useful for rebasing.
